### ISSUE TITLE
[Design] New encryption standard

### ISSUE THREAD

**Issue by r4mmer** [HathorNetwork/hathor-wallet-lib#637]
- Feature Name: wallet_encryption_standard
- Start Date: 2024-04-19
- Author: André Carneiro <andre.carneiro@hathor.network>

# Summary
[summary]: #summary

Update the wallet-lib's encryption standard, increasing security and offering more flexibility to the wallets.

# Motivation
[motivation]: #motivation

After a npm audit alert about some insecurities on PBKDF2 usage we decided to investigate our usage and this is meant to propose some improvements to how we store sensitive data.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Current standard

The current encrypted data follows this model.
```typescript
interface IEncryptedData {
  data: string; // encrypted data
  hash: string; // PBKDF2 of the secret
  salt: string; // salt, iterations and pbkdf2Hasher are used in the hash above
  iterations: number;
  pbkdf2Hasher: string;
}
```

We use PBKDF2 with the default of 1000 iterations to save the "hash" of the secret and use AES encryption to calculate the data using the secret.

This means `data = AES.encrypt(value, secret)` and `hash = PBKDF2(secret, config)` this means that we can use the `hash` to validate the user input before actually decrypting the data or to just validate that the user has inputted the correct secret.

This has some issues since PBKDF2 current security standard should use 1.3 million iterations with the sha1 hasher (`pbkdf2Hasher`) or 600 thousand iterations with the sha256 hasher.
Also AES encryption with the raw secret is less secure than if we used a key derivation function to generate the AES secret.

## Proposed standard

AES encryption is pretty safe if a proper encryption key is used, so we will use the user password to generate the encryption key instead of using the password directly.
This means that we also cannot save the hash of the password, since it makes it easier to derive the key.

```typescript
interface IEncryptedData {
  data: string; // encrypted secret+data
  hash: string; // sha256 of secret
  kdfConfig: IKdfConfig; // KDF config
}

interface IPBKDF2Config {
  kdf: 'PBKDF2';
  salt: string;
  iterations: string;
  hasher: string;
}

interface KdfConfig = IPBKDF2Config | ... // can include other KDFs
```

AES decryption may return some garbage output if the wrong password is used we will add 32 bytes to the encrypted data.
These 32 bytes will be used as a secret to check that the decryption process worked, by saving the sha256 hash of this secret we can check that the decrypted data matches.

This process may not work, but the 32 bytes of decrypted garbage would need to match the secret, this has about $2^(-256)$ chance in happening.

```typescript
// > encryption
// generate a key from the password and kdf config
const key = generateKey(password, kdfConfig);
// generate a random secret
const secret = generateSecret();
// encrypt the secret + value
const data = AES.encrypt(secret+value, key);
// save the sha256 hash of the random secret
const hash = sha256(secret);


// > decryption
// generate a key using the password from user input (slow process)
const key = generateKey(password, kdfConfig);
// decrypt the data using AES
const decryptedData = AES.decrypt(data, key);
// get the random secret from the decrypted data (length must be higher than 32 bytes)
const possibleSecret = decryptedData[:32];
if (sha256(possibleSecret) === hash) {
  // decryption successfull
  const value = decryptedData[32:];
} else {
  // decryption error: wrong password
}
```

As stated before the storage will hold the encrypted data, the hash of a random secret and the KDF configuration.
With this information we can safely decrypt and detect that the correct password was used.
We will not loose any functionality for our wallets while achieving a higher security for the stored data.

The other benefit of this proposal is the flexibility to choose another KDF, change its parameters to have increased security or update the defaults while not having to migrate all wallets to the new default.

## Migration process

To avoid migration issues we should leave the old standard implemented as the legacy and rename the `IEncryptedData` to `ILegacyEncryptedData`, we would then be able to check the stored data and determine which method of decryption to use.

All new wallets will be created with the new standard but old wallets may use the old standard.
The wallet should present the option to migrate the wallet (will require the user to input the pin and password) but if the user does not want to migrate he will be able.

**Comment by r4mmer** [comment]
reviewers: @pedroferreira1 @msbrogli

**Comment by pedroferreira1** [comment]
> For migrations from versions v0.26.0 and previous, we cannot safely migrate to the new standard since the encrypted data has changed, so we can keep the migration to the legacy standard and show the migration warning so the user can manually migrate to the new standard.

I didn't understand this. This migration will happen after the user upgrades his wallet. Versions before v0.26.0 are the ones without the new storage scheme, then we can migrate the storage as we already do, and have the migration warning, can't we?

---

The `checkPassword` and `decryptData` have duplicated code. For the final version we must improve this.

--- 

> Each wallet may configure the default number of iterations but for mobile we should keep the 10 thousand and for desktop we should change to 100 thousand iterations.

How long does it take to encrypt/decrypt the data with this amount of iterations?

--- 

```
function isLegacyEncryptedData(data: IEncryptedData): data is ILegacyEncryptedData {
  // Just check for a key not present on INewEncryptedData
  return 'iterations' in data;
}
```

We won't store the number of iterations in the storage anymore?

**Comment by r4mmer** [comment]
> I didn't understand this. This migration will happen after the user upgrades his wallet. Versions before v0.26.0 are the ones without the new storage scheme, then we can migrate the storage as we already do, and have the migration warning, can't we?

Yes, that is what I mean by "keep the migration to the legacy standard and show the migration warning so the user can manually migrate to the new standard".
In this context the "legacy standard" is how the new storage scheme handles encrypted access data.

---

> How long does it take to encrypt/decrypt the data with this amount of iterations?

It depends on the hardware, on my personal machine with 100k iterations I got 600ms and on the phone with 10k iterations I got 2s.
We can adjust this and under "Future possibilities > Guessing the best KDF config for the hardware" I explain how we can have a more personalized iteration count.

---

>We won't store the number of iterations in the storage anymore?

Yes we do, but inside the `kdfConfig`.
The iteration count is a PBKDF2 configuration (hence `IPBKDF2Config`), other KDFs will have other configuration parameters.
For instance Argon2 can have how many threads and the amount of memory each call can consume, since these are KDF specific i decided to move the KDF config to the `kdfConfig` and make adding other KDFs easier in the future.

**Comment by pedroferreira1** [comment]
For me this is approved
